network/wsNetwork_test.go

```diff
@@ -3949,6 +3950,123 @@ func TestTryConnectEarlyWrite(t *testing.T) {
 	assert.Equal(t, uint64(1), netA.peers[0].miMessageCount)
 }
 
+// Test functionality that allows a node to discard a block response that it did not request or that arrived too late.
+// Both cases are tested here by having A send unexpected, late responses to nodes B and C respectively.
+func TestDiscardUnrequestedBlockResponse(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	netA := makeTestWebsocketNode(t, testWebsocketLogNameOption{"netA"})
+	netA.config.GossipFanout = 1
+
+	netB := makeTestWebsocketNode(t, testWebsocketLogNameOption{"netB"})
+	netB.config.GossipFanout = 1
+
+	netC := makeTestWebsocketNode(t, testWebsocketLogNameOption{"netC"})
+	netC.config.GossipFanout = 1
+
+	netA.Start()
+	defer netA.Stop()
+	netB.Start()
+	defer netB.Stop()
+
+	addrB, ok := netB.Address()
+	require.True(t, ok)
+	gossipB, err := netB.addrToGossipAddr(addrB)
+	require.NoError(t, err)
+
+	netA.wg.Add(1)
+	netA.tryConnect(addrB, gossipB)
+	time.Sleep(250 * time.Millisecond)
+	require.Equal(t, 1, len(netA.peers))
```

Comment:
Instead of a blanket sleep here, you could do Eventually, like you do below

Reply:
Calling len on the netA.peers would cause a data race between calling the len and network writing it when it establishes a peer I believe

Reply:
I would very much like a race safe way of confirming that a peer has connected for testing purposes though if you or anyone else has ideas =) I don't like the unconditional sleep here either

Reply:
what about something like
```suggestion
	ctx, cancel = context.WithTimeout(context.Background(), 250 * time.Millisecond)
	defer cancel()
	initialValue := atomic.LoadInt32(&netA.peersChangeCounter)
	netA.tryConnect(addrB, gossipB)
	for {
		select {
		case <- ctx.Done():
			assert.NotEqual (atomic.LoadInt32(&netA.peersChangeCounter), initialValue, "250 ms and no peer attached")	
			break
		default:
			time.Sleep(time.Millisecond * 50)
			if atomic.LoadInt32(&netA.peersChangeCounter) != initialValue {
				break
			}
		}
	}

```

Reply:
That works thanks.